Support --web-header option for flutter run

aran opened the pull request:
Adds support for a new --web-header option to flutter run.

Creates a workaround for https://github.com/flutter/flutter/issues/127902

This PR allows adding additional headers for the flutter run web server. This is useful to add headers like Cross-Origin-Embedder-Policy and Cross-Origin-Opener-Policy without the use of a proxy server. These headers are required enable advanced web features. This approach provides flexibility to the developer to make use of the feature as they see fit and is backward-compatible. One tradeoff is that it increases the surface area to support for future changes to the flutter web server.

https://github.com/flutter/flutter/issues/127902 is not fully addressed by this change. The solution for that task will be more opinionated. This PR creates a general-purpose workaround for anyone who needs a solution sooner while the bigger solution is developed.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

aran committed:
Support --web-header option for flutter run

Adds support for a new --web-header option to flutter run.

Fixes https://github.com/flutter/flutter/issues/127902

aran committed:
Move parsing of --web-header option to flutter_command, follow the http rules, and add a test.

michaelnew commented:
For whatever it's worth, I just tested this out with `--web-header=Cross-Origin-Opener-Policy=same-origin --web-header=Cross-Origin-Embedder-Policy=require-corp`, and that's allowed me to use sqlite3 wasm with the OPFS to read/write to a sqlite database in the browser. Previously I had to build for web and then host with my own server that set those headers, but now I can run with the debugger attached.

Huge win for me, so thank you for this PR. I hope it gets merged soon!

andrewkolos reviewed (changes requested):
Thanks for the PR!

The intent of this looks good to me from a CLI tool perspective. Please ~~update [this test](https://github.com/flutter/flutter/blob/daea6e00bfc0452cbd3b1c46f3dd2d41a05da0a7/packages/flutter_tools/test/commands.shard/hermetic/run_test.dart#L1080)~~ (edit: I realized that test is specifically for flags, not options) and add a test for a case where the user provides an invalid header and a tool exit gets thrown. I will be adding another reviewer shortly

andrewkolos commented:
@eyebrowsoffire Can you review this from a web perspective and make sure this is an appropriate resolution for #127902 ? Also, if you happen to know of an existing integration test that this feature could be added to, please let us know; it would be nice to have this under test end-to-end.

aran committed:
Make HttpRegex class private and move comments closer to where they are relevant

aran committed:
Add tests for web-header parsing

aran commented on packages/flutter_tools/lib/src/runner/flutter_command.dart:
```diff
@@ -218,6 +229,14 @@ abstract class FlutterCommand extends Command<void> {
+            'as headers to all responses. Multiple headers can be passed by '
+            'repeating "--web-header" multiple times.',
+      valueHelp: 'X-Custom-Header=header-value',
+      splitCommas: false,
```
It probably won't make a difference in the real world, but unless I'm mistaken commas and equal-signs are allowed in http header values. One of the test cases I'm adding will discriminate this edge case.

eyebrowsoffire reviewed (approved):
Looks good to me. As discussed on Discord, I don't believe this is a full resolution to https://github.com/flutter/flutter/issues/127902 and we should not close that issue based on just this PR. I think we need more specific and turnkey handling of the CORS headers needed for `crossOriginIsolated` rather than a generic mechanism like this. Nonetheless, this change seems good to me and will be generally useful for end users.

andrewkolos committed:
fix and move tests

Moves tests into their own "--web-header" group. Also updates the invalid input tests to check against the exception message to make sure we are throwing a ToolExit with the correct message. After doing this, I noticed tests were passing despite the thrown error message having nothing to do with the --web-header option, so I fixed them.

andrewkolos committed:
Merge branch 'master' into headers

andrewkolos committed:
finish tweaking test names

andrewkolos commented:
Thanks @eyebrowsoffire!

@aran, I've pushed some changes which simply move and rename the tests to be more consistent with the general pattern found throughout the rest of the CLI tool test code.

I also made some of the `throwsToolExit` checks more strict by adding error message checks. Upon doing this, I noticed some of the `ToolExit`s being thrown weren't the ones we were actually testing for, so I tweaked some of the test setup to resolve this. Feel free to make additional tweaks.

andrewkolos reviewed (approved):
LGTM. I will merge this after CI checks pass and @aran has had ample time to respond to my latest commits should they choose to do so

aran commented:
Great collaborating with you on this, thanks for all the help and for taking this change.

andrewkolos commented:
> Great collaborating with you on this, thanks for all the help and for taking this change.

Likewise; thanks for your efforts and contribution!
